src/state_machine.rs

```diff
@@ -0,0 +1,1062 @@
+            }))
+    }
+
+    /// Return the `Strategy` used to construct mutators  for this
+    /// `StateMachine`.
+    pub fn mutation_strategy(&self) -> BoxedStrategy<Box<dyn Mutator<S>>> {
+        self.mutation_strategy.clone()
+    }
+
+    /// Return the function used for testing invariants.
+    pub fn invariants(&self) -> Arc<dyn Fn (&mut S) -> TestCaseResult> {
+        Arc::clone(&self.invariants)
+    }
+
+    /// Create a strategy for test cases using this state machine.
+    ///
+    /// `state` gives a strategy used for generating the initial state.
+    ///
+    /// `size` specifies the possible number of steps to take.
+    pub fn test_case
+        (&self, state: impl Strategy<Value = S>, size: impl Into<SizeRange>)
+         -> impl Strategy<Value = StateMachineTestCase<S>>
+    {
+        let clone = self.clone();
+
+        (state, collection::vec(self.mutation_strategy(), size))
+            .prop_map(move |(state, mutations)| StateMachineTestCase::new(
+                clone.clone(), state, mutations))
+    }
+}
+
+impl<S : fmt::Debug> fmt::Debug for StateMachine<S> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("StateMachine")
+            .field("mutation_strategy", &self.mutation_strategy)
+            .field("invariants", &"<function>")
+            .finish()
+    }
+}
+
+/// A single test-case of a state machine.
+///
+/// The only useful operation is `.run()`.
+#[must_use = "StateMachineTestCase does nothing unless .run() is called"]
+pub struct StateMachineTestCase<S : fmt::Debug> {
+    state: S,
+    mutations: Vec<Box<dyn Mutator<S>>>,
+    machine: StateMachine<S>,
+    trace: bool,
+}
+
+impl<S : fmt::Debug> fmt::Debug for StateMachineTestCase<S> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("StateMachineTestCase")
+            .field("state", &self.state)
+            .field("mutations", &self.mutations)
+            .field("machine", &"<elided for clarity>")
+            .field("trace", &self.trace)
+            .finish()
+    }
+}
+
+impl<S : fmt::Debug> StateMachineTestCase<S> {
+    fn new(machine: StateMachine<S>,
+           state: S,
+           mutations: Vec<Box<dyn Mutator<S>>>) -> Self {
+        StateMachineTestCase { state, mutations, machine, trace: false }
+    }
+
+    /// Enable tracing on this test case.
+    ///
+    /// Tracing causes verbose logging to be emitted about each mutation as
+    /// they are performed.
+    pub fn trace(mut self) -> Self {
+        self.trace = true;
+        self
+    }
+
+    /// Run this test case.
+    ///
+    /// Returns `Ok` with the final internal state if successful. On failure,
+    /// returns `Err` with the appropriate error, or panics if user code
+    /// triggered a panic.
+    pub fn run(mut self) -> Result<S, TestCaseError> {
+        if self.trace {
+            println!("initial state: {:?}", self.state);
+            println!("check initial invariants");
```

Comment:
A thought for later: It might be nice to change from using `println` to use some more principled logging facility that users can control. I think the quickcheck crate does this. But this is out of scope for this PR.